### PR TITLE
[algo] Add SPO (Single-stream Policy Optimization) recipe implementation

### PR DESCRIPTION
**Pull request by dzh19990407**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

This PR implements the Single-stream Policy Optimization proposed by paper https://arxiv.org/abs/2509.13232.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/volcengine/verl/pull/2775
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

**Comment by vermouth1992** [comment]
Could you pin the verl commit in your readme?

File: recipe/spo/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+| **SPO** | **84.0** | **76.5** | **46.9** | **64.0** | **47.5** | **63.8** |
+| **Gain** | **+0.7** | **+4.4** | **+1.3** | **+7.3** | **+3.3** | **+3.4** |
+
+## 🎯 Quick Start
+
+### Basic Training
+```bash
+# Enable SPO training mode
+export SPO_ENABLE=True
+export SPO_OFFLINE_VALUES="/path/to/offline/values.json"
```
Comment (hustnn) [inline recipe/spo/README.md]: what is the purpose of this file?

Comment (zhongwen-xu) [reply recipe/spo/README.md]: see Appendix A

Comment (dzh19990407) [reply recipe/spo/README.md]: This file is an offline value estimate (Appendix A), I have added a link to huggingface in the README.

**Comment by dzh19990407** [comment]
> Could you pin the verl commit in your readme?

I have update it in the readme file.

**Comment by dzh19990407** [comment]
@wuxibin89 @vermouth1992 @tongyx361 @PeterSH6 
Hi team,

Thanks for all the great feedback! I have updated the code based on the review comments and pushed the changes.

Please take a quick look when you have a chance. Thanks!

**Comment by zhongwen-xu** [comment]
@hustnn @wuxibin89 @vermouth1992 @tongyx361 @PeterSH6 

Hey folks, can you review this PR?

**Comment by hustnn** [comment]
Hi @zhongwen-xu ，I am not in the verl team.  I am interested in your work and testing it in my case (a coding model). If it working, I will let you know.  

Do I need to first run it to get offline values before training?

**Comment by zhongwen-xu** [comment]
Hi @hustnn, 

Thanks for your interests. For your question, please refer to Appendix E in https://arxiv.org/abs/2509.13232, especially Figure 6 (c) for "Offline Initialization Ablation".

Short answer: the offline initialization helps for early training steps, in the long run, with and without the initializaon are similar. 

My recommendation is to get the offline values before training though, a good v_hat estimation with the offline sampling is relatively cheap, same cost as what people do for filtering prompts, and they can be shared across multiple experiment runs.

Happy to answer any further questions by email!

**Comment by zsgvivo** [comment]
Hi @zhongwen-xu ,Thanks for your great work on SPO. I'm testing your implementation using the training scripts provided in this PR (`recipe/spo/run_spo.sh`). I've noticed that in both SPO and GRPO setup, the model use tools at the beginning, but as training progresses, it gradually stops calling them and instead generates direct answers. Is this behavior expected?  
截屏2025-10-18 15 23 06

**Comment by zhongwen-xu** [comment]
Hi @zsgvivo

Thanks for your interests in our work! 
The reason we decided to close this PR was that we initially thought a quick PR with the original ReTool tool protocol might work with Qwen3 8B without SFT, so we submitted the PR before the holiday. However, during the long holiday, we did a verification run with this PR implementation, it turned out it did't work, similar to what you just reported. Now we suspected that the ReTool protocol requires SFT first, rather than directly RL from it.

We are working on a **full replication** of what we had in the paper where the tool call protocol is with  
\`\`\`python  
 the generated code here
\`\`\`

, rather than the JSON function calls as it was in ReTool implementation. (Note that we actually present all the core code in this PR so if you can modify the tool call protocol as what I just said, it should just work, without waiting for our implementation.)

Stay tuned!

**Comment by longls777** [comment]
Hi @zhongwen-xu and SPO authors,

Thanks a lot for sharing your work and the initial SPO implementation in this PR.

From your last comment, I understand that this PR version didn’t pass your verification run and that you’re working on a full replication with the updated tool-calling protocol (instead of the original ReTool JSON function calls). I’m very interested in reproducing the SPO results and applying the method to my own use cases.

May I ask if there is currently a validated/recommended implementation of SPO available (e.g., a newer branch, updated scripts, or a different repo) that you would suggest users follow?
If there is any WIP code, configs, or guidance you’re comfortable sharing—even if it’s not fully polished yet—it would be extremely helpful.

Thanks again for the great work, and really appreciate any pointers!

**Comment by dzh19990407** [comment]
@longls777 SPO has been merged into https://github.com/verl-project/verl-recipe

**Comment by longls777** [comment]
> 

Thanks so much for sharing this implementation - really appreciate your help!
